linux/src/timer_linux.c

```diff
@@ -0,0 +1,224 @@
+
+#include <time.h>
+
+#include "azure_macro_utils/macro_utils.h"
+
+#include "azure_c_logging/xlogging.h"
+#include "azure_c_pal/gballoc_ll.h"
+#include "azure_c_pal/gballoc_ll_redirect.h"
+
+#include "azure_c_pal/timer.h"
+
+typedef struct TIMER_HANDLE_DATA_TAG
+{
+    struct timespec start_time;
+}TIMER_HANDLE_DATA;
+
+TIMER_HANDLE timer_create_new(void)
+{
+    TIMER_HANDLE result;
+
+    /* Codes_SRS_TIMER_LINUX_01_001: [ timer_create_new shall allocate memory for a new timer. ]*/
+    result = malloc(sizeof(TIMER_HANDLE_DATA));
+    if (result == NULL)
+    {
+        /* Codes_SRS_TIMER_LINUX_01_003: [ If any error occurs, timer_create_new shall return NULL. ]*/
+        LogError("failure in malloc");
+        /*return as is*/
+    }
+    else
+    {
+        struct timespec tp;
+        /* Codes_SRS_TIMER_01_001: [ timer_create_new shall create a new timer and on success return a non-NULL handle to it. ]*/
+        /* Codes_SRS_TIMER_LINUX_01_002: [ timer_create_new shall call clock_gettime with CLOCK_MONOTONIC to obtain the initial timer value. ]*/
+        if (clock_gettime(CLOCK_MONOTONIC, &result->start_time) != 0)
+        {
+            /* Codes_SRS_TIMER_LINUX_01_003: [ If any error occurs, timer_create_new shall return NULL. ]*/
+            LogError("clock_gettime failed");
+        }
+        else
+        {
+            goto all_ok;
+        }
+
+        free(result);
+    }
+
+    result = NULL;
+
+all_ok:
+    return result;
+}
+
+void timer_destroy(TIMER_HANDLE timer)
+{
+    if (timer == NULL)
+    {
+        /* Codes_SRS_TIMER_01_002: [ If timer is NULL, timer_destroy shall return. ]*/
+        /* Codes_SRS_TIMER_LINUX_01_004: [ If timer is NULL, timer_destroy shall return. ]*/
+        LogError("Invalid arguments: TIMER_HANDLE timer=%p", timer);
+    }
+    else
+    {
+        /* Codes_SRS_TIMER_01_003: [ Otherwise, timer_destroy shall free the memory associated with timer. ]*/
+        /* Codes_SRS_TIMER_LINUX_01_005: [ Otherwise, timer_destroy shall free the memory associated with timer. ]*/
+        free(timer);
+    }
+}
+
+int timer_start(TIMER_HANDLE timer)
+{
+    int result;
+
+    if (timer == NULL)
+    {
+        /* Codes_SRS_TIMER_01_004: [ If timer is NULL, timer_start shall fail and return a non-zero value. ]*/
+        /* Codes_SRS_TIMER_LINUX_01_007: [ If timer is NULL, timer_start shall fail and return a non-zero value. ]*/
+        LogError("Invalid arguments: TIMER_HANDLE timer=%p", timer);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        /* Codes_SRS_TIMER_01_005: [ Otherwise, timer_start shall record the start time (used for computing the elapsed time). ]*/
+        /* Codes_SRS_TIMER_LINUX_01_006: [ timer_start shall call clock_gettime with CLOCK_MONOTONIC to obtain the start timer value. ]*/
+        if (clock_gettime(CLOCK_MONOTONIC, &timer->start_time) != 0)
+        {
+            LogError("clock_gettime failed");
+            result = MU_FAILURE;
```

Comment:
Looks like this should have a requirement for failing

Reply:
There was one, it is now tagged here.

---
In reply to: [512808976](https://github.com/Azure/azure-c-pal/pull/48#discussion_r512808976) [](ancestors = 512808976)